### DISCUSSION TITLE
How can I get a Testcontainers handle on the application container launched by a @QuarkusIntegrationTest?

### DISCUSSION THREAD

**Discussion by albers** [quarkusio/quarkus#30494] [Q&A] [answered]
**Context:**
When I launch Maven integration tests via

```shell
mvn -DskipITs=false -Dquarkus.container-image.build verify
```

and a `@QuarkusIntegrationTest` is present, an application image is built and an application container launched for test purposes behind the scenes.

**Question:**
Is there a way to inject or otherwise get access to this application container **as a Testcontainers object** in the test class or a `QuarkusTestResourceLifecycleManager`?

[a comment by an automated account is not shown]

**Comment by geoand** [comment]
We don't currently have a way to do this, but I would like to hear more about the use case you have in mind.

**Reply by albers** [reply]
Thanks @geoand.

My builds run on a build agent container created by a Docker service inside a Docker Swarm. This causes networking issues during integration tests.

In a QuarkusIntegrationTest, I want the MP Rest Client of my Quarkus app container to access the Wiremock instance created on my build agent. The automatically launched test container cannot access the build agent container because they do not share a common network.

I see two possible solutions for this problem:
- [Testcontainers magic](https://www.testcontainers.org/features/networking/#exposing-host-ports-to-the-container) - this is why I opened this discussion.
- Temporarily add the build agent container to the dev services network for the duration of the integration tests by means of a  QuarkusTestResourceLifecycleManager.

**Reply by geoand** [reply]
So if I understand correctly, if you know the network on which the application container will be launched, you can work around the issue?

**Reply by albers** [reply]
Yes, that is true.

I just got some code working that retrieves the network ID from the `DevServicesContext` and then uses Testcontainers and dockerjava to search the build agent container (inspired by [this fragment](https://github.com/quarkusio/quarkus/blob/bad8aa2ce25036b768a4683da715b079b2feae9b/extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ContainerLocator.java#L31-L35)). I then add the build container to the dev services network. This is done in the `QuarkusTestResourceLifecycleManager` that manages my Wiremock instance.

I was hoping for a more elegant solution.

**Reply by geoand** [reply]
What kind of more elegant generalized solution were you thinking of?

**Reply by albers** [reply]
I see two helpful additions.

- The container ID or a Testcontainers instance of the current test container could be an injectable resource in `@QuarkusIntegrationTest`s.
- The `DevServicesContext` could give me access to container IDs or Testcontainers instances of containers created by dev services or an integration test.

**Reply by geoand** [reply]
The "problem" is that `@QuarkusIntegrationTest` is useful for more than just container testing, so maybe providing some utility classes would make more sense.

**Reply by albers** [reply]
For the application container, a conceptually clean solution could be to inject an `Optional`. This would clearly state that the container is not available under all circumstances.

**Reply by geoand** [reply]
Yeah, that's also an option.

Mind opening an issue asking for this feature?

**Reply by albers** [reply]
I still wonder whether this issue is of general interest.

**Reply by geoand** [reply]
It probably would be useful for some people
